fog/recovery_db_iface/src/types.rs

```diff
@@ -5,8 +5,82 @@
+    /// This should be the latest block that existed before we published it (or, a block close to but before that)
+    pub start_block: u64,
+    /// The largest pubkey expiry value that we have ever published for this key.
+    /// If less than start_block, it means we have never published this key.
+    pub pubkey_expiry: u64,
+    /// Whether this key is retiring / retired.
+    /// When a key is retired, we stop publishing reports about it.
+    pub retired: bool,
+}
+
+impl IngressPublicKeyStatus {
+    /// Whether a block index lies between start_block and pubkey_expiry for this key.
+    /// If the key is not retired yet, we assume that pubkey_expiry may be increasing,
+    /// so we only check if self.start_block <= block_index in that case.
+    ///
+    /// If it does, then the block index potentially contains TxOut's which had
+    /// fog hints encrypted using this ingress public key, and fog needs to scan
+    /// this block with this ingress key, or declare it a missed block.
+    pub fn covers_block_index(&self, block_index: u64) -> bool {
+        self.start_block <= block_index && (!self.retired || block_index < self.pubkey_expiry)
+    }
+}
+
+/// Information returned after attempting to add block data to the database.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct AddBlockDataStatus {
+    /// Indicates that the block we tried to add has already been scanned using this ingress key,
+    /// and didn't need to be scanned again.
+    ///
+    /// If this value is true, then no data was added to the database.
+    pub block_already_scanned_with_this_key: bool,
+}
+
+/// IngressPublicKeyRecord
+///
+/// This is returned by get_ingress_public_key_records, and augments the PublicKeyStatus so that
+/// the last_block_scanned is also returned, as well as the public key bytes themselves.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct IngressPublicKeyRecord {
+    /// The ingress public key this data refers to
+    pub key: CompressedRistrettoPublic,
+    /// The status of the key
+    pub status: IngressPublicKeyStatus,
+    /// The last block scanned by this key.
+    /// This is inherently racy since other partcipants may be writing concurrently with us, but this
+    /// number is a lower bound.
+    pub last_scanned_block: Option<u64>,
+}
+
+impl IngressPublicKeyRecord {
+    /// The next block index that needs to be scanned with this key.
+    ///
+    /// This is one of:
+    /// - last_scanned_block + 1
+    /// - start_block if last_scanned_block is None
+    /// - None, we're not actually on the hook for that block, per self.covers_block_index
+    pub fn next_needed_block_index(&self) -> Option<u64> {
+        let candidate = self
+            .last_scanned_block
+            .map(|x| x + 1)
+            .unwrap_or(self.status.start_block);
+        if self.status.covers_block_index(candidate) {
+            Some(candidate)
+        } else {
+            None
+        }
+    }
+}
```

Comment:
in the future, we should use this function in ingest instead of duping this logic

it's not used in this revision though I think, I ended up not using it. we could remove it for now